### PR TITLE
refactor(versioning/cargo): enable strict null

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes:

- Fix strict null errors.

## Context:

- Ref: #12869 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/versioning/cargo/index.ts [hunk 1]
```diff
@@ -55,19 +55,27 @@ function npm2cargo(input: string): string {
 }
 
 const isLessThanRange = (version: string, range: string): boolean =>
-  npm.isLessThanRange(version, cargo2npm(range));
+  !!npm.isLessThanRange?.(version, cargo2npm(range));
```
Comment (viceice) [inline lib/versioning/cargo/index.ts:58]: Are you sure we need this change?

File: lib/versioning/cargo/index.ts [hunk 2]
```diff
@@ -96,7 +104,7 @@ function getNewValue({
     currentVersion,
     newVersion,
   });
-  let newCargo = npm2cargo(newSemver);
+  let newCargo = npm2cargo(newSemver ?? '');
```
Comment (viceice) [inline lib/versioning/cargo/index.ts:107]: This looks wrong to me.

File: lib/versioning/cargo/index.ts (hunk 1 above)
Comment (RahulGautamSingh) [reply lib/versioning/cargo/index.ts:58]: Yes, checked it using `yarn null-check`

- ? is necessary as per `VersioningApi` in `types.ts` and it makes the overall type (boolean|undefined) so,
- !! converts falsy values to boolean(false).

File: lib/versioning/cargo/index.ts (hunk 2 above)
Comment (RahulGautamSingh) [reply lib/versioning/cargo/index.ts:107]: Error: newSemver has type string | null while `npm2cargo` allows only string.

Solution: Fallback to an empty string when newSemver is null. An empty string should work just fine cause:
https://github.com/renovatebot/renovate/blob/2b78fa575e145edecf0c491320f2776ba72e9bab/lib/versioning/cargo/index.ts#L37-L40

In the case of falsy input, it returns the same input which is then used as `newCargo`.
```js
  let newCargo = npm2cargo(newSemver);
  // istanbul ignore if
  if (!newCargo) {
    logger.info(
      { currentValue, newSemver },
      'Could not get cargo version from semver'
    );
    return currentValue;
  }
  ```

**Comment by RahulGautamSingh** [comment]
🟢@viceice Ready for review.

**Comment by viceice** [comment]
> 🟢@viceice Ready for review.

I know, please don't make to much noise:

https://github.com/renovatebot/renovate/blob/main/.github/contributing.md#re-requesting-a-review

File: lib/versioning/cargo/index.ts (hunk 2 above)
Comment (viceice) [reply lib/versioning/cargo/index.ts:107]:
```suggestion
  let newCargo = newSemver ? npm2cargo(newSemver): null;
```

This looks cleaner to me. Maybe `npm2cargo` should be widen to allow null values

**Review by viceice — COMMENTED** [review]
see comment

**Comment by RahulGautamSingh** [comment]
> > 🟢@viceice Ready for review.
> 
> I know, please don't make to much noise:
> 
> https://github.com/renovatebot/renovate/blob/main/.github/contributing.md#re-requesting-a-review

Sorry, won't happen again. 
This PR needed to be merged before I could push some changes to other branches so, I got a bit impatient. My bad

**Comment by viceice** [comment]
There is just a small thing left. So waiting for your changes. 🙃

**Comment by RahulGautamSingh** [comment]
The file `lib/util/cache/package/redit.ts` throws an error while checking for null errors. This was not happening in previous tests so maybe some new update has introduced this one. Since this file is not part of the current PR I have tried excluding it from strict-null-checks.

**Review by viceice — APPROVED** [review]

**Pull request merged**